homeassistant/components/pterodactyl/sensor.py

```diff
@@ -0,0 +1,208 @@
+"""Sensor platform of the Pterodactyl integration."""
+
+from __future__ import annotations
+
+from collections.abc import Callable
+from dataclasses import dataclass
+from typing import Any
+
+from homeassistant.components.sensor import SensorEntity, SensorEntityDescription
+from homeassistant.const import (
+    PERCENTAGE,
+    EntityCategory,
+    UnitOfInformation,
+    UnitOfTime,
+)
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity_platform import AddConfigEntryEntitiesCallback
+from homeassistant.helpers.typing import StateType
+from homeassistant.util.unit_conversion import DurationConverter, InformationConverter
+
+from .coordinator import PterodactylConfigEntry, PterodactylCoordinator, PterodactylData
+from .entity import PterodactylEntity
+
+KEY_CPU_UTILIZATION = "cpu_utilization"
+KEY_MEMORY_UTILIZATION = "memory_utilization"
+KEY_DISK_UTILIZATION = "disk_utilization"
+KEY_NETWORK_INBOUND = "network_inbound"
+KEY_NETWORK_OUTBOUND = "network_outbound"
+KEY_UPTIME = "uptime"
+
+# Coordinator is used to centralize the data updates.
+PARALLEL_UPDATES = 0
+
+
+@dataclass(frozen=True, kw_only=True)
+class PterodactylSensorEntityDescription(SensorEntityDescription):
+    """Class describing Pterodactyl sensor entities."""
+
+    value_fn: Callable[[PterodactylData], StateType]
+    attributes_fn: Callable[[PterodactylData], dict[str, Any]] | None
+
+
+def get_extra_state_attributes_cpu(
+    data: PterodactylData,
+) -> dict[str, list[str]]:
+    """Return CPU limit as extra state attribute."""
+    extra_state_attributes: dict[str, Any] = {}
+
+    extra_state_attributes["cpu_limit"] = data.cpu_limit
+
+    return extra_state_attributes
+
+
+def get_extra_state_attributes_memory(
+    data: PterodactylData,
+) -> dict[str, list[str]]:
+    """Return memory usage and limit as extra state attributes."""
+    extra_state_attributes: dict[str, Any] = {}
+
+    extra_state_attributes["memory_usage"] = round(
+        InformationConverter.convert(
+            data.memory_usage, UnitOfInformation.BYTES, UnitOfInformation.MEGABYTES
+        )
+    )
+    extra_state_attributes["memory_limit"] = round(
+        InformationConverter.convert(
+            data.memory_limit, UnitOfInformation.BYTES, UnitOfInformation.MEGABYTES
+        )
```

Comment:
Idem

Reply:
So we should have all 3 entities? Usage (percentage), usage (absolute) and limit?

Reply:
Yes, but I'm not sure what to think about the percentage yet as we ideally don't have calculated values